app/controllers/resources_controller.rb

```diff
@@ -59,20 +59,48 @@ def check_permission
     privilege = params[:privilege]
     raise ArgumentError, "privilege" unless privilege
 
-    result = assumed_role.allowed_to?(privilege, resource)
+    begin
+      result = assumed_role.allowed_to?(privilege, resource)
+
+      audit_success(privilege, result)
+    rescue => e
+      audit_failure(privilege, e)
+
+      raise
+    end
 
+    head(result ? :no_content : :not_found)
+  end
+
+  def audit_success(privilege, result)
     Audit.logger.log(
       Audit::Event::Check.new(
         user: current_user,
         client_ip: request.ip,
-        resource: resource,
+        resource_id: resource.id,
         privilege: privilege,
-        role: assumed_role,
+        role_id: assumed_role.id,
         operation: "check",
         success: result
       )
     )
+  end
 
-    head(result ? :no_content : :not_found)
+  def audit_failure(privilege, err)
+    resource_id = "#{params[:account]}:#{params[:kind]}:#{params[:identifier]}"
+    role_id = params[:role]
+
+    Audit.logger.log(
+      Audit::Event::Check.new(
+        user: current_user,
+        client_ip: request.ip,
+        resource_id: resource_id,
```

Comment:
Maybe i'm missing something, in the value, why in here it's `resource_id` and not `resource.id` (same for role_id)

Reply:
Still don't understand, if the role or resource does not exists why do we audit it like it is exists?

Reply:
Because we need to audit the checked role/resource so will be able to read the audit and understand what the user tried to do(i think that it gives us the ability to tell if the user just misspelled the role id or trying to do something else). any way thats what was expected from this feature

Reply:
@netacoral Please notice that we will have a reliability issue here, as success messages are guaranteed to hold a real/valid resource - while on failure we take what we get from query param...

auditor must be aware to this line.